server/server.go

```diff
@@ -1,31 +1,54 @@
 package main
 
 import (
+	"errors"
 	"io"
 	"log"
 	"net/http"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
 )
 
-// server does not use a multiplexer, as we only care about http method, query params and request body
 type server struct {
-	db     *database
-	server *http.Server
+	db                   *database
+	mux                  *http.ServeMux
+	requestCounterMetric prometheus.Counter
 }
 
-func (s *server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	if r.URL.Path != "/" {
-		w.WriteHeader(http.StatusBadRequest)
-		return
+func (s *server) routes() {
+	s.mux.HandleFunc("/db", s.metricsMiddleware(s.requestLoggerMiddleware(s.handleDB())))
+	s.registerMetrics()
+
+	s.mux.HandleFunc("/*", s.handleBadPath())
+}
+
+func (s *server) handleDB() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		key := r.URL.Query().Get("key")
+		switch r.Method {
+		case http.MethodDelete:
+			s.handleDelete(w, key)
+		case http.MethodGet:
+			s.handleGet(w, key)
+		case http.MethodPut:
+			s.handlePut(w, r, key)
+		default:
+			s.handleNotImplemented(w)
+		}
 	}
-	key := r.URL.Query().Get("key")
-	switch r.Method {
-	case http.MethodGet:
-		s.handleGet(w, key)
-	case http.MethodPut:
-		s.handlePut(w, r, key)
-	default:
-		handleNotImplemented(w)
+}
+
+const ti = 1
+
+func (s *server) handleDelete(w http.ResponseWriter, key string) {
+	err := s.db.delete(key)
```

Comment:
I know you know what this error means, but if you have a typed error you should use it